openstf/model/objective.py

```diff
@@ -78,49 +149,49 @@ def __call__(self, trial: optuna.trial.FrozenTrial) -> float:
         # Configure evals for early stopping
         eval_set = [(train_x, train_y), (valid_x, valid_y)]
 
-        model = ModelCreator.create_model(self.model_type)
-
-        params = self.get_params(trial)
+        # get the parameters used in this trial
+        param = self.get_trial_parameters(self.model_params, trial)
 
-        model.set_params(**params)
+        # create the specific pruning callback
+        pruning_callback = self.pruning_function(trial, **self.args_eval)
 
-        eval_metric = self.eval_metric
+        # insert parameters into model
+        self.model.set_params(**param)
 
-        # See https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.XGBoostPruningCallback.html
-        pruning_callback = optuna.integration.XGBoostPruningCallback(
-            # Use validation_0 to prune on the train set
-            # Use validation_1 to prune on the validation set
-            trial,
-            observation_key=f"validation_1-{eval_metric}",
-        )
         # validation_0 and validation_1 are available
-        model.fit(
+        self.model.fit(
             train_x,
             train_y,
             eval_set=eval_set,
             early_stopping_rounds=EARLY_STOPPING_ROUNDS,
             verbose=self.verbose,
-            eval_metric=eval_metric,
+            eval_metric=self.eval_metric,
             callbacks=[pruning_callback],
         )
 
-        forecast_y = model.predict(test_x)
+        forecast_y = self.model.predict(test_x)
 
-        return mae(test_y, forecast_y)
+        try:
```

Comment:
Perhaps we can make this a little nicer. A good rule is always to fail as early as possible, so in this case the constructor should already fail if the right evaluation metric is not available. I would suggest something like this:

```python
# at the top of the objective
from openstf.metrics import metrics

# in the RegressorObjective construction
def __init__():
    self.eval_metric_function = metrics.get_metric_eval_function(eval_metric)

# and in the __call__ method
def __call__():
    return self.eval_metric_function(test_y, forecast_y)

# in openstf.metric.metrics add a get_metric_evaluation_function
def get_metric_eval_function(metric_name):
    evaluation_function = {
        "bias": bias,
        "nsme": nsme,
        "mae": mae,
        "r_mae": r_mae,
        "frac_in_stdev": frac_in_stdev,
        "r_mae_highest": r_mae_highest,
        "r_mne_highest": r_mne_highest,
        "r_mpe_highest": r_mpe_highest,
        "r_mae_lowest": r_mae_lowest,
        "skill_score": skill_score,
        "skill_score_positive_peaks": skill_score_positive_peaks,
        "franks_skill_score": franks_skill_score,
        "franks_skill_score_peaks": franks_skill_score_peaks,
    }.get(metric_name, None)

    if evaluation_function is None:
        raise KeyError(f"Unknown evaluation metric function {metric_name}")

    return evaluation_function

```